### DISCUSSION TITLE
Integrate custom CSS/JS plugin into Quire

### DISCUSSION THREAD

**Discussion by yl5682** [thegetty/quire#41] [Development]
I am hoping to integrate [a third-party plugin](http://vcg.isti.cnr.it/rti/webviewer.php) but when I was referencing the JS/CSS files, Quire would report that files are not found (404). I doubled checked the file paths and URLs but the 404 error persists. 

Please see attachments:
<img width="886" alt="Screen Shot 2021-03-01 at 2 23 14 PM" src="https://user-images.githubusercontent.com/7308403/109547905-fa213f80-7a99-11eb-9f54-09c1b7467f69.png">

<img width="617" alt="Screen Shot 2021-03-01 at 2 25 54 PM" src="https://user-images.githubusercontent.com/7308403/109547990-145b1d80-7a9a-11eb-9503-b0d360d519f4.png">

<img width="616" alt="Screen Shot 2021-03-01 at 2 26 00 PM" src="https://user-images.githubusercontent.com/7308403/109547992-14f3b400-7a9a-11eb-8a49-7d3712be243f.png">

<img width="1158" alt="Screen Shot 2021-03-01 at 2 27 13 PM" src="https://user-images.githubusercontent.com/7308403/109548347-8af81b00-7a9a-11eb-8107-1aeb26f7f178.png">

<img width="450" alt="Screen Shot 2021-03-01 at 2 29 41 PM" src="https://user-images.githubusercontent.com/7308403/109548374-95b2b000-7a9a-11eb-9334-9ff060cd520e.png">

I feel like there is a config that determines what paths/files are allowed to be rendered in the Quire server. Any input would be helpful. Thank you

**Comment by geealbers** [comment]
Hi Yang, that's peculiar. You have everything set up correctly. You made the template changes directly to `themes/default/layouts/partials/head.html`? And you put the CSS files themselves in the `static/css` directory of your project, or where?

**Reply by yl5682** [reply]
Hey Greg! Thanks for your pointer! I think I might have just missed one file path or two. Now I have it working and I am documenting the steps in a separate comment under this thread. Hope it helps others. 

In the meantime, I don't know if this type of overriding is the best practice or not (esp. considering conflicts due to future Quire updates). If you have any advice regarding customization that'd be helpful. I don't want to mislead others in the community.

Thanks!

**Reply by geealbers** [reply]
Glad it's working, and thanks so much for posting your full solution here for others!

We haven't developed a good path for customization with external libraries like this yet. To avoid the future updates issue you mentioned, I think ideally we'd build some logic in to a couple of the partials based on config parameters or other sources external to the partials so that they could be updated programatically instead of manually overridden. 

For instance, in `layouts/partials/head.html` we could add logic that would look for any and all files placed in the `static/css` and `static/js` directories and automatically add links to them.

**Reply by yl5682** [reply]
That'd be great. Thanks for being so active in the community helping users like us out!

**Comment by yl5682** [comment]
Here is my update:

**Insert the HTML snippet to create the instance**

```html
<!--
 Added Web RTI Viewer code here
 /themes/default/layouts/entry/single.html
-->

<!-- Web RIT Viewer -->
  <div id="viewerContainer">
    <script type="text/javascript">
      createRtiViewer("viewerContainer", "webrti", 900, 600); 
    </script>
  </div>
<!-- End of Web RIT Viewer -->
```

**Include external assets in the head.html**

```html
{{- $jquery := "webViewer/css/ui-lightness/jquery-ui-1.10.3.custom.css" | relURL -}}
{{- $webrtiviewerCSS := "webViewer/css/webrtiviewer.css" | relURL -}}
{{- $webrtiviewerCSS := "webViewer/css/webrtiviewer.css" | relURL -}}
{{- $spiderglJS := "webViewer/spidergl/spidergl.js" | relURL -}}
{{- $multiresJS := "webViewer/spidergl/multires.js" | relURL -}}

<!-- WebRTI Viewer -->
  <link type="text/css" href="{{ $jquery }}" rel="stylesheet">
  <link type="text/css" href="{{ $webrtiviewerCSS }}" rel="stylesheet">

	<!-- Seems I can also use this style to include -->
	<!-- But I will need the leading '/' -->
  <script type="text/javascript" src="/webViewer/js/jquery-1.11.3.min.js"></script>
  <script type="text/javascript" src="/webViewer/js/jquery-ui.min.js"></script>
  <script type="text/javascript" src="/webViewer/js/pep.min.js"></script>
  
  <script type="text/javascript" src="{{ $spiderglJS }}"></script>
  <script type="text/javascript" src="{{ $multiresJS }}"></script>
<!-- WebRTI Viewer -->
```

**Place webViewer directly inside static folder**
<img width="618" alt="Screen Shot 2021-03-16 at 1 15 04 PM" src="https://user-images.githubusercontent.com/7308403/111352289-65524080-865a-11eb-8277-60465eb21d05.png">


**Place catalogue files directly inside static folder**
<img width="616" alt="Screen Shot 2021-03-16 at 1 15 46 PM" src="https://user-images.githubusercontent.com/7308403/111352308-697e5e00-865a-11eb-8e0c-98cbcb9a6986.png">


Note that the webViewer JavaScript files will be looking for files inside of the catalogue folder. Ensure that the file paths match.

End result:

<img width="1756" alt="Screen Shot 2021-03-16 at 1 21 51 PM" src="https://user-images.githubusercontent.com/7308403/111352487-99c5fc80-865a-11eb-9399-3eab09f0d617.png">
